lightgbm/src/main/scala/com/microsoft/ml/spark/lightgbm/dataset/DatasetAggregator.scala

```diff
@@ -0,0 +1,488 @@
+    featuresChunkedArray.delete()
+  }
+}
+
+class BaseAggregatedColumns extends AggregatedColumns {
+  var labelsArray: FloatSwigArray = _
+  var weightArrayOpt: Option[FloatSwigArray] = None
+  var initScoreArrayOpt: Option[DoubleSwigArray] = None
+  var groupColumnValuesArray: Array[Row] = _
+
+  /**
+    * Variables for knowing how large full array should be allocated to
+    */
+  @volatile var rowCount = 0L
+  @volatile var initScoreCount = 0L
+
+  var numCols = 0L
+}
+
+trait BaseDenseAggregatedColumns extends BaseAggregatedColumns {
+  var featuresArray: DoubleSwigArray = _
+
+  def incrementCount(rowCount: Long,
+                     initScoreCount: Long): Unit
+
+  def addRows(labelsChunkedArray: FloatChunkedArray,
+              weightChunkedArrayOpt: Option[FloatChunkedArray],
+              initScoreChunkedArrayOpt: Option[DoubleChunkedArray],
+              featuresChunkedArray: DoubleChunkedArray,
+              groupColumnValues: ListBuffer[Row],
+              numCols: Int): Unit
+}
+
+class DenseAggregatedColumns(chunkSize: Int) extends BaseDenseAggregatedColumns {
+  def incrementCount(rowCount: Long,
+                     initScoreCount: Long): Unit = {
+    this.rowCount += rowCount
+    this.initScoreCount += initScoreCount
+  }
+
+  /** Adds the rows to the internal data structure.
+    * @param labelsChunkedArray The column of label values.
+    * @param weightChunkedArrayOpt The optional column of weights, if specified.
+    * @param initScoreChunkedArrayOpt The optional column of initial scores, if specified.
+    * @param featuresChunkedArray The features vector.
+    * @param groupColumnValues The column of group values, if in ranking scenario.
+    */
+  def addRows(labelsChunkedArray: FloatChunkedArray,
+              weightChunkedArrayOpt: Option[FloatChunkedArray],
+              initScoreChunkedArrayOpt: Option[DoubleChunkedArray],
+              featuresChunkedArray: DoubleChunkedArray,
+              groupColumnValues: ListBuffer[Row],
+              numCols: Int): Unit = {
+    initializeRows(weightChunkedArrayOpt, initScoreChunkedArrayOpt, numCols)
+    // Coalesce to main arrays passed to dataset create
+    labelsChunkedArray.coalesceTo(this.labelsArray)
+    weightChunkedArrayOpt.foreach(_.coalesceTo(this.weightArrayOpt.get))
+    initScoreChunkedArrayOpt.foreach(_.coalesceTo(this.initScoreArrayOpt.get))
+    featuresChunkedArray.coalesceTo(this.featuresArray)
+    for (index <- 0 until groupColumnValues.length) {
+      groupColumnValuesArray(index) = groupColumnValues(index)
+    }
+  }
+
+  private def initializeRows(weightChunkedArrayOpt: Option[FloatChunkedArray],
+                             initScoreChunkedArrayOpt: Option[DoubleChunkedArray],
+                             numCols: Int): Unit = {
+    this.numCols = numCols
+    labelsArray = new FloatSwigArray(this.rowCount)
+    weightChunkedArrayOpt.foreach(weightChunkedArray => {
+      weightArrayOpt = Some(new FloatSwigArray(this.rowCount))
```

Comment:
It may be cleaner to write this as:
```scala
weightArrayOpt = weightChunkedArrayOpt.map(_ => new FloatSwigArray(this.rowCount))
```

Reply:
done!

Reply:
Not a big deal though, feel free to ignore this one.